_config/security.yml

```diff
@@ -26,9 +26,14 @@ SilverStripe\Core\Injector\Injector:
     properties:
       Middlewares:
         AuthenticationMiddleware: '%$SilverStripe\Security\AuthenticationMiddleware'
+        BasicAuthMiddleware: '%$SilverStripe\Security\BasicAuthMiddleware'
   SilverStripe\Security\AuthenticationMiddleware:
     properties:
       AuthenticationHandler: '%$SilverStripe\Security\AuthenticationHandler'
+  SilverStripe\Security\BasicAuthMiddleware:
+    properties:
+      ExcludeURLPatterns: []
+      IncludeURLPatterns: []
```

Comment:
Just curious, what if I want my project to explicitly include a path that is excluded by a module, or vice-versa? Should we enforce an associative array here instead?


_security.yml_
```
SilverStripe\Security\BasicAuthMiddleware:
  properties:
    ExcludeURLPatterns:
        'my-path': true
        'my-other-path': 'ADMIN'
```

_mymodule.yml_
```
SilverStripe\Security\BasicAuthMiddleware:
  properties:
    ExcludeURLPatterns:
        'my-path': false
```

Reply:
ok, so we use the bool flag for "enabled" right? And if it's a string, it's a permission check (except for exclude, which doesn't make sense to check for permissions).

Reply:
Hm, it's a bit ambiguous if both key and value are strings; Someone might try to put the pattern as the value, and use a named key instead (like we do for extensions sometimes).

Reply:
Ok, let's combine into one value; Keys are patterns, and value are:
 - bool true - force on (no permissions)
 - string - force on (special permission)
 - bool false - force exclude
 - null - default to global setting (controlled via yml)

Reply:
```php
/**
     * URL Patterns for basic auth. Keys are the Regexp string to match, and the key can
     * be one of the below:
     *  - true (bool) - Enabled for this url
     *  - false (bool) - Disabled for this url
     *  - Any string / array - Enabled for this url, and require the given string as a permission code
     *  - null (default) - Calls BasicAuth::protect_site_if_necessary(), which falls back to config setting
     *
     * @see CanonicalURLMiddleware
     * @var array
     */
    protected $urlPatterns = [];
```

Reply:
Brilliant. Love this change.